subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	const struct bt_data *current_data = &stream->ad[stream->ad_index].data[stream->data_index];
+
+	const bool is_first_data = (stream->ad_index == 0) && (stream->data_index == 0);
+	const bool is_last_data = (stream->ad_index == stream->ad_len - 1) &&
+				  (stream->data_index == stream->ad[stream->ad_index].len - 1);
+
+	if (stream->current_data_offset == 0) {
+		/* Start reading out a new bt_data structure */
+
+		size_t max_data_to_be_read = current_data->data_len + 2;
+
+		buf[0] = current_data->data_len + 1;
+		buf[1] = current_data->type;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
+
+	} else {
+
+		/* Read the remaining data */
+		ssize_t max_data_to_be_read = current_data->data_len - stream->current_data_offset;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	uint8_t operation;
+
+	if (is_first_data && is_last_data &&
+	    stream->current_data_offset == current_data->data_len) {
+		operation = BT_HCI_LE_EXT_ADV_OP_COMPLETE_DATA;
+	} else if (is_first_data) {
+		operation = BT_HCI_LE_EXT_ADV_OP_FIRST_FRAG;
+	} else if (is_last_data && stream->current_data_offset == current_data->data_len) {
+		operation = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+	} else {
+		operation = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+	}
+
+	/* Increment counters */
+	if (stream->current_data_offset == current_data->data_len) {
```

Comment:
If this is true, does that mean it is the last fragment? 

Or is there a case where it isn't the last fragment, but this is still true?

Reply:
It means it was the last fragment of the current advertising data structure. That is not the same as BT_HCI_LE_EXT_ADV_OP_LAST_FRAG as there may be more advertising data structures coming.

Reply:
Right. This is due to the fact that we send the set data command for each LTV structure, right? (Which arguable is still a horribly inefficient way of doing that, but that's outside the scope of this PR). 

But correct me if I am wrong then: 
`is_first_data` seems to be true if it is very first data, and not just the first part of one of the LTV structures. `is_last_data` is however not the last data, but rather the last LTV structure. 

`is_last_data` may then be a bit ill-named if it indeed not the last data as you mention in https://github.com/zephyrproject-rtos/zephyr/pull/41537/files#r778169407

Or did I misunderstand you?

Reply:
ìs_first_data` and `is_last_data` refer to the LTV structures. Maybe these should be rephrased then? I tried to use the existing terminology.

Reply:
if `is_first_data` is the first LTV, and `is_last_data` also refer to the last LTV, I'd just replace `data` with `ltv` or `ad`.